concert/experiments/imaging.py

```diff
@@ -60,3 +63,555 @@ def tomo_max_speed(frame_width, frame_rate):
+
+    async def _get_radio_position(self):
+        return self._radio_position
+
+    async def _get_flat_position(self):
+        return self._flat_position
+
+    async def _set_num_flats(self, n):
+        self._num_flats = int(n)
+
+    async def _set_num_darks(self, n):
+        self._num_darks = int(n)
+
+    async def _set_num_projections(self, n):
+        self._num_projections = int(n)
+
+    async def _set_flat_position(self, position):
+        self._flat_position = position
+
+    async def _set_radio_position(self, position):
+        self._radio_position = position
+
+    async def _prepare_flats(self):
+        await self._flat_motor.set_position(await self.get_flat_position())
+        await self.start_sample_exposure()
+
+    async def _prepare_darks(self):
+        await self.stop_sample_exposure()
+
+    async def _prepare_radios(self):
+        await self._flat_motor.set_position(await self.get_radio_position())
+        await self.start_sample_exposure()
+
+    async def _finish_radios(self):
+        """
+        Function that is called after all frames are acquired. It will be called only once.
+        """
+        if self._finished:
+            return
+        await self.stop_sample_exposure()
+        self._finished = True
+
+    @background
+    async def run(self):
+        self._finished = False
+        await super().run()
+
+    async def _take_radios(self):
+        try:
+            await self._prepare_radios()
+            async for frame in self._produce_frames(self._num_projections):
+                yield frame
+        finally:
+            await self._finish_radios()
+
+    async def _take_flats(self):
+        try:
+            await self._prepare_flats()
+            async for frame in self._produce_frames(self._num_flats):
+                yield frame
+        finally:
+            await self._prepare_darks()
+
+    async def _take_darks(self):
+        try:
+            await self._prepare_darks()
+            async for frame in self._produce_frames(self._num_darks):
+                yield frame
+        finally:
+            await self._prepare_darks()
+
+    async def _produce_frames(self, number, after_acquisition=None):
+        """
+        Generator of frames
+
+        :param number: Number of frames that are generated
+        :param after_acquisition: function that is called after all frames are acquired (but maybe not yet downloaded).
+        Could be None or a Future.
+        :return:
+        """
+        await self._camera.set_trigger_source("AUTO")
+        async with self._camera.recording():
+            for i in range(int(number)):
+                yield await self._camera.grab()
+        if after_acquisition is not None:
+            await after_acquisition()
+
+    async def start_sample_exposure(self):
+        raise NotImplementedError()
+
+    async def stop_sample_exposure(self):
+        raise NotImplementedError()
+
+
+class SteppedTomography(Radiography):
```

Comment:
Shouldn't we call this just `Tomography`. `ContinuousTomography` is for me the opposite of stepped and here it inherits...

Reply:
To make it really correct, from the inheritance, nameing etc. I could do
![imaging](https://user-images.githubusercontent.com/8747063/146633126-bae3dde0-aa45-4ce5-90fa-36c95d1d54c9.png)

Reply:
And then also add SteppedSpiralTomo, where I need to think a little bit to not add to much code duplication.

Reply:
Yes, but if `SteppedTomography` turns out to be just empty then you can just omit it.

Reply:
Tomography will be almost emplty. Only the rot-motor and some parameters will be introduced. Then the thw tomo-implementations will implement the prepare_radios and take_radios.
I think this is the most clear and most educational solution.